JDK-8352869

eme64 opened the pull request:
TODO

galderz commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,76 @@
      */
-    public static void checkEQ(Object a, Object b) {
-        checkEQ(a, b, "");
+    public static void checkEQ(Object a, Object b, boolean isFloatCheckWithRawBits, boolean isCheckWithArbitraryClasses) {
```
Just a suggestion. One boolean might be ok, but once you start adding 2 booleans it seems like a bit of a code smell to me. Do you envision more options being added? I would personally create a `VerifyOptions` record with the boolean flags options and pass that in.

eme64 replied:
I'll experiment with a `VerifyOptions`, thanks for the suggestion!

galderz commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +113,182 @@ private static void checkEQ(Object a, Object b, String context) {
+     */
+    private void checkEQimpl(float a, float b, String field, Object aParent, Object bParent) {
+        if (isFloatEQ(a, b)) {
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch. check raw: " + isFloatCheckWithRawBits);
```
Would using a text block here make it more readable?

eme64 replied:
@galderz What do you mean by a text block? String Templates would be nice, but we don't have them. Do you mean I should use `String.format`? But there is always so tricky to know what is going to be formatted to where...

galderz commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +113,182 @@ private static void checkEQ(Object a, Object b, String context) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch for " + context);
+    private void checkEQimpl(double a, double b, String field, Object aParent, Object bParent) {
+        if (isDoubleEQ(a, b)) {
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch. check raw: " + isFloatCheckWithRawBits);
```
Same text block comment here

galderz commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +320,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+    }
+
+    private void print(Object a, Object b, String field, Object aParent, Object bParent) {
+        System.err.println("  aParent: " + aParent);
```
Text block?

galderz commented on test/hotspot/jtreg/testlibrary_tests/verify/tests/TestVerify.java:
```diff
@@ -400,20 +462,233 @@ public static void testRecursive() {
         } catch (VerifyException e) {}
     }
+
+    public static void checkNE(Object a, Object b) {
```
If `checkEQ` lives in Verify, wouldn't it make sense to also have `checkNE` there? Seems like the natural place making the API symmetric.

eme64 replied:
To be honest, I don't see the need for it being symmetric, i.e. for the API to have a `checkNE`. This `checkNE` method is just a convenience function to check that an exception is thrown if I expect it to.

eme64 commented:
@galderz Thanks for the comments! I think I addressed / answered all of them. Could you please have another look? :)

galderz commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +146,182 @@ private static void checkEQ(Object a, Object b, String context) {
+     */
+    private void checkEQimpl(float a, float b, String field, Object aParent, Object bParent) {
+        if (isFloatEQ(a, b)) {
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch. check raw: " + verifyOptions.isFloatCheckWithRawBits);
```
I meant using JEP 378 text blocks, e.g.

```suggestion
System.err.printf("""
    ERROR: Verify.checkEQ failed: value mismatch. check raw: %b
      Values: %.1f vs %.1f
      Raw:    %d vs %d
    """, isFloatCheckWithRawBits, a, b, Float.floatToRawIntBits(a), Float.floatToRawIntBits(b));
```

eme64 replied:
I see.

That has advantages and disadvantages.
Advantage: You can more easily see the "skeleton" of the test.
Disadvantage: Mapping the "holes" and the "values" is annoying, you basically have to count through each position. Plus it may end up being more lines.

Best would really be String Templates.

I asked @chhagedorn , he does not have an opinion either way.

Personally, I prefer my way, where you can easily see what values go where directly. But this is probably a taste question.

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +146,182 @@ private static void checkEQ(Object a, Object b, String context) {
+                    return;
+                }
+
+                if (verifyOptions.isCheckWithArbitraryClasses) {
+                    checkEQArbitraryClasses(a, b);
+                    return;
+                } else {
+                    System.err.println("ERROR: Verify.checkEQ failed: type not supported: " + ca.getName());
+                    print(a, b, field, aParent, bParent);
+                    throw new VerifyException("Object type not supported: " + ca.getName() + " -- did you mean to 'enableCheckWithArbitraryClasses'?");
+                }
```
What's the reason behind throwing instead of just comparing two arbitrary objects by default? If a user calls `Verify.checkEQ()` and sees this exception, I would guess he then just passes the additional option and we have the same result. But maybe I'm missing something.

eme64 replied:
Good question. I think my reasoning was that comparing arbitrary classes requires reflection. And that is rather slow. So by default it would be good if that feature is not enabled, so the user tries to avoid it, and is aware when they enable it explicitly.

But if you think that is not useful, I can remove the feature.

@chhagedorn what do you think?

chhagedorn replied:
I think the intention to let the user double check is good. I'm not sure though if the user is really aware of the potential slow down without diving deeper into the implementation. All they know is that `checkEQ` somehow does not support their some objects but there is a simple workaround to still use it. So, the real question is: How many users will then consider doing something different when facing this exception and not just enable it anyway? I guess enabling is probably the most natural thing to do.

Given that, I would probably just drop this. It would also simplify the API usage in the following way:

We would only have checks with NaNs being all equals and comparing raw bits (i.e. NaNs not equal). Then you could offer `checkEQ()` (default) and `checkRawBitsEQ()` or something like that. Then users do not need to worry about creating and passing in an `Options`.

What do you think about these suggestions?

What we could do either way at the `checkEQ()` API method: Describe the potential slow down with reflection when not using certain classes.

chhagedorn reviewed (commented):
Nice extensions! Some initial comments.

eme64 commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,109 @@
 
```
done :)

eme64 commented:
@chhagedorn Thanks for the suggestions and questions! I think I addressed them all :)

chhagedorn reviewed (commented):
I'll have a closer look at the code later again :-)

eme64 commented:
@chhagedorn Ok, I refactored it. I'm now always comparing arbitrary classes. And `checkEQWithRawBits` does the comparison with raw bits, no `Options` required any more. Added a comment about reflection making things slow.

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,72 @@
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.ArrayList;
```
Seems unused and can be removed.
```suggestion
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,72 @@
  */
 public final class Verify {
+    private final boolean isFloatCheckWithRawBits;
+    private final HashMap<Object, Object> a2b = new HashMap<>();
+    private final HashMap<Object, Object> b2a = new HashMap<>();
```
Can you add a comment here what `a2b` and `b2a` means?

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,72 @@
 
     /**
-     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     * Verify the content of two Objects, possibly recursively.
```
Maybe add:
```suggestion
     * Verify the contents of two Objects on a raw bit level, possibly recursively.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,72 @@
 
     /**
-     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     * Verify the content of two Objects, possibly recursively.
```
```suggestion
     * Verify the contents of two Objects, possibly recursively.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
-            case double[]  x -> checkEQimpl(x, (double[])b,                context);
-            case MemorySegment x -> checkEQimpl(x, (MemorySegment) b,      context);
+            case Object[]  x -> checkEQimpl(x, (Object[])b,                 field, aParent, bParent);
+            case Byte      x -> checkEQimpl(x, ((Byte)b).byteValue(),       field, aParent, bParent);
```
Can't you just pass `(Byte) b` to rely on auto unboxing instead?

eme64 replied:
You are right, simplified it :)

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
         if (ca != cb) {
```
Only seen this in my IDE: `ca` and `cb` should be `Class<?>` instead of the raw `Class` since `getClass()` returns a `Class<?>` (cannot make a suggestion since it's hidden here).

eme64 replied:
fixed

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
-                print(a, "a " + context);
-                print(b, "b " + context);
-                throw new VerifyException("Object array type not supported: " + ca.getName());
+                if (ca.getName().startsWith("jdk.incubator.vector") && ca.getName().contains("Vector")) {
```
Might be worth to extract this case to own methods and structure it like this to reduce the size of the method:
```
if (vectorClass()) {
    checkEQForVectorAPIClass();
} else {
    checkEQdispatch();
}
```

eme64 replied:
Refactored it :)

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+                        m.setAccessible(true);
+                        va = m.invoke(a);
+                        vb = m.invoke(b);
+                    } catch (NoSuchMethodException e) {
+                        throw new RuntimeException("Could not invoke toArray on " + ca.getName(), e);
+                    } catch (IllegalAccessException e) {
+                        throw new RuntimeException("Could not invoke toArray on " + ca.getName(), e);
+                    } catch (InvocationTargetException e) {
+                        throw new RuntimeException("Could not invoke toArray on " + ca.getName(), e);
+                    }
```
You can merge them:
```suggestion
                    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
                        throw new RuntimeException("Could not invoke toArray on " + ca.getName(), e);
                    }
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+    private void checkEQimpl(char a, char b, String field, Object aParent, Object bParent) {
         if (a != b) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + (int)a + " vs " + (int)b + " for " + context);
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + (int)a + " vs " + (int)b);
```
Why do you need an upcast here? Same for `short`.

eme64 replied:
Look at this ;)
```
jshell> char a = 66;
a ==> 'B'

jshell> System.out.println("a: " + a);
a: B

jshell> System.out.println("a: " + (int)a);
a: 66
```

But I can remove the casts for `short`.

eme64 replied:
Added a comment as well.

chhagedorn replied:
Right, that makes sense for the `char` case. But good that we could remove it for the `short` case.

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+     * But the Java spec allows us to return different bits for a NaN, which allows swapping the inputs
+     * of an add or mul (NaN1 * NaN2 does not have same bits as NaN2 * NaN1, because the multiplication
+     * of two NaN values should always return the first of the two).
+     * Hence, by default, we pick the non-raw coparison: we verify that we have the same bit
```
```suggestion
     * Hence, by default, we pick the non-raw comparison: we verify that we have the same bit
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+    }
+
+    /**
+     * See comments for "isFloatEQ".
```
We don't have Javadocs for the private methods but it could still help when navigating in the IDE to directly jump the the method when clicking on it (suggested same below for other places):
```suggestion
     * See comments for {@link #isFloatEQ}.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+    }
+
+    /**
+     * Check that two floats are equal according to "isFloatEQ".
```
```suggestion
     * Check that two floats are equal according to {@link #isFloatEQ}.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+     */
+    private void checkEQimpl(float a, float b, String field, Object aParent, Object bParent) {
+        if (isFloatEQ(a, b)) {
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch. check raw: " + isFloatCheckWithRawBits);
```
Just noticed this now (there are other places as well): Since we now have `Verify.checkEQ()` and `Verify.checkEQWithRawBits()`, it would improve the readability if we reported which method was used. It could be done with something like that (pseudo code):
```Java
System.err.println("ERROR: Verify.checkEQ" + withRawBitsString() + " failed: value mismatch.

String withRawBitsString() {
    return isFloatCheckWithRawBits ? "WithRawBits" : "";
}
```

eme64 replied:
Boah. That is really going to bloat the code, don't you think?
The exception that is thrown will already give you the complete stack trace, including which methods were called. Is that not good enough?

chhagedorn replied:
Hm, it could indeed be a little bit more complicated when you are deep down in a recursion. My thought was that it could be misleading when a test is using a mix of `verifyEQ()` and `verifyEQWithRawBits()` and you only read `verifyEQ` failed. You could be start looking at the wrong check even though the stack trace would have guided you to the correct place. Maybe we can just update "Verify.checkEQ" into something more generic like "Equality matching failed" and we're good. What do you think?

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+        if (isFloatEQ(a, b)) {
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch. check raw: " + isFloatCheckWithRawBits);
+            System.err.println("  Values: " + a + " vs " + b);
+            System.err.println("  Raw:    " + Float.floatToRawIntBits(a) + " vs " + Float.floatToRawIntBits(b));
```
Do we always want to dump the raw bits even when `isFloatCheckWithRawBits` is false? I guess it does not hurt.

eme64 replied:
Yes, I want that. It can help if there are different `NaN` encodings. Or if we somehow reinterpreted integer values as floats. It's been useful for me in the past :)

chhagedorn replied:
Sounds good, let's leave it in then!

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -180,15 +287,16 @@ private static void checkEQimpl(double a, double b, String context) {
      * Verify that the content of two MemorySegments is identical. Note: we do not check the
```
```suggestion
     * Verify that the contents of two MemorySegments are identical. Note: we do not check the
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
         throw new VerifyException("MemorySegment value mismatch.");
     }
 
+    /**
+     * Verify that the content of two MemorySegments is identical. Note: we do not check the
+     * backing type, only the size and content.
+     */
```
Probably a copy-paste error. Should be updated for exceptions.

eme64 replied:
Good catch, updated it!

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+    }
+
     /**
      * Verify that the content of two byte arrays is identical.
```
```suggestion
     * Verify that the contents of two byte arrays are identical.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
     }
 
     /**
      * Verify that the content of two char arrays is identical.
```
```suggestion
     * Verify that the contents of two char arrays are identical.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
     }
 
     /**
      * Verify that the content of two short arrays is identical.
```
```suggestion
     * Verify that the contents of two short arrays are identical.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
     }
 
     /**
      * Verify that the content of two int arrays is identical.
```
```suggestion
     * Verify that the contents of two int arrays are identical.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
     }
 
     /**
      * Verify that the content of two long arrays is identical.
```
```suggestion
     * Verify that the contents of two long arrays are identical.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+    }
+
+    /**
+     * Verify that the content of two boolean arrays is identical.
```
```suggestion
     * Verify that the contents of two boolean arrays are identical.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
     }
 
     /**
      * Verify that the content of two Object arrays is identical, recursively:
```
```suggestion
     * Verify that the contents of two Object arrays are identical, recursively:
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
-        if (a == null) {
-            System.err.println("  " + context + ": null");
+    private void checkEQArbitraryClasses(Object a, Object b) {
+        Class c = a.getClass();
```
```suggestion
        Class<?> c = a.getClass();
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+        Class c = a.getClass();
+        while (c != Object.class) {
+            for (Field field : c.getDeclaredFields()) {
+                Object va = null;
+                Object vb = null;
```
`null` can be omitted:
```suggestion
                Object va;
                Object vb;
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+        // Hence, we cannot use the mapping below. We test these boxed primitive types by value anyway,
+        // and they are no recursive structures, so there is no point in optimizing here anyway.
+        switch(a) {
+            case Boolean x -> { return false; }
+            case Byte x -> { return false; }
+            case Short x -> { return false; }
+            case Character x -> { return false; }
+            case Integer x -> { return false; }
+            case Long x -> { return false; }
+            case Float x -> { return false; }
+            case Double x -> { return false; }
```
I think the convention is to us `_` when they are ignored. You can then also merge them:
```suggestion
            case Boolean _, 
                 Byte _, 
                 Short _, 
                 Character _, 
                 Integer _, 
                 Long _, 
                 Float _, 
                 Double _ -> { return false; }
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+        Object bPrevious = a2b.get(a);
+        Object aPrevious = b2a.get(b);
+        if (aPrevious == null && bPrevious == null) {
+            // Record for next time.
```
Can you explain, maybe as comment at `checkAlreadyVisited()`, why we want to have these caches?

eme64 replied:
Added documentation :)

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -289,7 +514,7 @@ private static void printMemorySegment(MemorySegment a, String context) {
+    private void printMemorySegmentValue(MemorySegment a, long offset, int range) {
         long start = Long.max(offset - range, 0);
         long end   = Long.min(offset + range, a.byteSize());
         for (long i = start; i < end; i++) {
```
Nit below: You can replace `System.err.println("")` with `System.err.println()`.

eme64 replied:
done!

chhagedorn commented on test/hotspot/jtreg/testlibrary_tests/verify/examples/TestWithVectorAPI.java:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
```

chhagedorn reviewed (commented):
Thanks for the update! It's much easier to use and understand now I think.

I did a complete pass and left a lot of comments but mostly minor things. Overall I think this looks great!

eme64 commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch for " + context);
-            System.err.println("       Values: " + a + " vs " + b);
-            System.err.println("       Values: " + Float.floatToRawIntBits(a) + " vs " + Float.floatToRawIntBits(b));
+    private boolean isFloatEQ(float a, float b) {
```
Good catch! Flipped the condition :)

eme64 commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+
+    private void print(Object a, Object b, String field, Object aParent, Object bParent) {
+        System.err.println("  aParent: " + aParent);
+        System.err.println("  bParent: " + bParent);
```
I think it does not hurt to print `null` here. It makes the code a little simpler.

chhagedorn replied:
Okay, maybe we can print `<none>` in case of a null for more clarity?

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,80 @@
-     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     * Verify the contents of two Objects on a raw bit level, possibly recursively.
+     * Different NaN encodings are considered non-equal, since we compare
+     * floating number by their raw bits.
```
```suggestion
     * floating numbers by their raw bits.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,80 @@
     /**
-     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     * Verify the contents of two Objects, possibly recursively.
+     * Different NaN encodins are considered equal.
```
```suggestion
     * Different NaN encodings are considered equal.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -203,82 +302,239 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
     /**
-     * Verify that the content of two byte arrays is identical.
+     * Verify that two Exceptions have the same message. Messages are not always carried,
+     * they are often dropped to performance, and that is ok. But if both Exceptions have
+     * the message, we should compare them.
```
```suggestion
     * they are often dropped for performance reasons, and that is okay. But if both Exceptions
     * have the message, we should compare them.
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -203,82 +302,239 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
-    private static void print(Object a, String context) {
-        if (a == null) {
-            System.err.println("  " + context + ": null");
+    /**
+     * We do not want to import jdk.incubator.vector explicitly, because it would mean we would also have
+     * to add "--add-modules=jdk.incubator.vector" to the command-line of every test that uses the Verify
+     * class. So we hack this via reflection.
+     */
```
I think this background is only needed at `checkEQForVectorAPIClass()` (where you already have that comment). Here you can just describe what the code actually does or just drop the comment entirely since the method name is self-explanatory :-)

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -25,33 +25,80 @@
+    /**
+     * When comparing arbitrary classes recursively, we need to remember which
+     * pairs of objects {@code (a, b)} we have already visited. The maps
+     * {@link a2b} and {@link b2a} track these edges. Caching which pairs
```
I think it's fine to use `code` here since the Javadocs links to itself otherwise.
```suggestion
     * {@code a2b} and {@code b2a} track these edges. Caching which pairs
```

chhagedorn commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -203,82 +302,239 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+    /**
+     * When comparing arbitrary classes recursively, we need to remember which
+     * pairs of objects {@code (a, b)} we have already visited. The maps
+     * {@link a2b} and {@link b2a} track these edges. Caching which pairs
```
```suggestion
     * {@link #a2b} and {@link #b2a} track these edges. Caching which pairs
```

chhagedorn reviewed (commented):
Thanks for addressing all my comments and doing the updates! I have some more final comments but then I think it's good to go from my side!

eme64 commented:
@chhagedorn Thanks for having another look! I applied all your suggestions :)

chhagedorn reviewed (approved):
That looks good to me, thanks for bearing with me!

eme64 commented:
@galderz do you intend to review / approve this, or should I ask someone else?

turbanoff commented on test/hotspot/jtreg/testlibrary_tests/verify/tests/TestVerify.java:
```diff
@@ -400,20 +462,233 @@ public static void testRecursive() {
+        }
+    }
+
+    public static class H1  {
```
```suggestion
    public static class H1 {
```

turbanoff commented on test/hotspot/jtreg/compiler/lib/verify/Verify.java:
```diff
@@ -203,82 +302,234 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
+        // Boxed primitives are not guaranteed to be the same Object for the same primitive value.
+        // Hence, we cannot use the mapping below. We test these boxed primitive types by value anyway,
+        // and they are no recursive structures, so there is no point in optimizing here anyway.
+        switch(a) {
```
```suggestion
        switch (a) {
```

eme64 commented:
@turbanoff Thanks for the two whitespace fixes :)

galderz commented:
@eme64 No, I don't think my review counts that much on this one. I think you need someone to review it that is has more background on the history of this.

eme64 commented:
@galderz Christian already reviewed it, and generally it is ok for someone with deeper knowledge and someone with a little less familiarity to review. So far, on `Verify.java` the only other commit is https://github.com/openjdk/jdk/pull/22715, so there is not too much history ;)

But totally up to you, I'm thankful for the comments you already left, and they do not obligate to do a complete review :)

TobiHartmann reviewed (approved):
Looks good to me! I just found a few minor typos.

eme64 commented:
@chhagedorn Thanks!

/integrate
